presto-native-execution/presto_cpp/main/PeriodicTaskManager.cpp

```diff
@@ -52,106 +52,132 @@ PeriodicTaskManager::PeriodicTaskManager(
 
 void PeriodicTaskManager::start() {
-  // Add new functions here.
-
   // If executors are null, don't bother starting this task.
   if (driverCPUExecutor_ or httpExecutor_) {
-    scheduler_.addFunction(
-        [driverCPUExecutor = driverCPUExecutor_,
-         httpExecutor = httpExecutor_]() {
-          if (driverCPUExecutor) {
-            // Report the current queue size of the thread pool.
-            REPORT_ADD_STAT_VALUE(
-                kCounterDriverCPUExecutorQueueSize,
-                driverCPUExecutor->getTaskQueueSize());
+    scheduleExecutorStatsTask();
+  }
+  if (taskManager_) {
+    scheduleTaskStatsTask();
+    scheduleTaskCleanupTask();
+  }
+  if (memoryAllocator_) {
+    scheduleMemoryAllocatorStatsTask();
+  }
+  schedulePrestoExchangeSourceMemoryStatsTask();
+  if (asyncDataCache_) {
+    scheduleAsyncDataCacheStatsTask();
+  }
+  scheduleConnectorStatsTask();
+  scheduleOperatingSystemStatsTask();
 
-            // Report the latency between scheduling the task and its execution.
-            folly::stop_watch<std::chrono::milliseconds> timer;
-            driverCPUExecutor->add([timer = timer]() {
-              REPORT_ADD_STAT_VALUE(
-                  kCounterDriverCPUExecutorLatencyMs, timer.elapsed().count());
-            });
-          }
+  // This should be the last call in this method.
+  scheduler_.start();
+}
 
-          if (httpExecutor) {
-            // Report the latency between scheduling the task and its execution.
-            folly::stop_watch<std::chrono::milliseconds> timer;
-            httpExecutor->add([timer = timer]() {
-              REPORT_ADD_STAT_VALUE(
-                  kCounterHTTPExecutorLatencyMs, timer.elapsed().count());
-            });
-          }
-        },
-        std::chrono::microseconds{kTaskPeriodGlobalCounters},
-        "executor_counters");
-  }
+void PeriodicTaskManager::stop() {
+  scheduler_.cancelAllFunctionsAndWait();
+  scheduler_.shutdown();
+}
 
-  if (taskManager_) {
-    scheduler_.addFunction(
-        [taskManager = taskManager_]() {
-          // Report the number of tasks and drivers in the system.
-          size_t numTasks{0};
-          auto taskNumbers = taskManager->getTaskNumbers(numTasks);
-          REPORT_ADD_STAT_VALUE(kCounterNumTasks, taskManager->getNumTasks());
-          REPORT_ADD_STAT_VALUE(
-              kCounterNumTasksRunning,
-              taskNumbers[velox::exec::TaskState::kRunning]);
-          REPORT_ADD_STAT_VALUE(
-              kCounterNumTasksFinished,
-              taskNumbers[velox::exec::TaskState::kFinished]);
-          REPORT_ADD_STAT_VALUE(
-              kCounterNumTasksCancelled,
-              taskNumbers[velox::exec::TaskState::kCanceled]);
-          REPORT_ADD_STAT_VALUE(
-              kCounterNumTasksAborted,
-              taskNumbers[velox::exec::TaskState::kAborted]);
+void PeriodicTaskManager::scheduleExecutorStatsTask() {
+  scheduler_.addFunction(
```

Comment:
I assume there is no logic change in the stats report here?

Reply:
Not at all